raft/node.go

```diff
@@ -68,18 +73,34 @@ type Ready struct {
 
 	// Entries specifies entries to be saved to stable storage BEFORE
 	// Messages are sent.
+	//
+	// If async storage writes are enabled, this field does not need to be acted
+	// on immediately. It will be reflected in a MsgStorageAppend message in the
+	// Messages slice.
 	Entries []pb.Entry
 
 	// Snapshot specifies the snapshot to be saved to stable storage.
+	//
+	// If async storage writes are enabled, this field does not need to be acted
+	// on immediately. It will be reflected in a MsgStorageAppend message in the
+	// Messages slice.
 	Snapshot pb.Snapshot
 
 	// CommittedEntries specifies entries to be committed to a
 	// store/state-machine. These have previously been committed to stable
 	// store.
+	//
+	// If async storage writes are enabled, this field does not need to be acted
+	// on immediately. It will be reflected in a MsgStorageApply message in the
+	// Messages slice.
 	CommittedEntries []pb.Entry
 
-	// Messages specifies outbound messages to be sent AFTER Entries are
-	// committed to stable storage.
+	// Messages specifies outbound messages.
+	//
+	// If async storage writes are not enabled, these messages must be sent
+	// AFTER Entries are committed to stable storage. If async storage writes
+	// are enabled, they can be sent immediately.
+	//
 	// If it contains a MsgSnap message, the application MUST report back to raft
 	// when the snapshot has been received or has failed by calling ReportSnapshot.
 	Messages []pb.Message
```

Comment:
Good point. Switched "synchronous"/"asynchronous" with "durable"/"non-durable".

You are correct that the field is unused with async append. Instead, the cases where `MustSync` used to be false now fall into two separate categories:
- **no HardState.Commit update**: in this case, there is no `MsgStorageAppend` message
- **only HardState.Commit update**: in this case, there is a `MsgStorageAppend` message, but it has no `Response` messages. This is nice because it means that logic that handles `MsgStorageAppend` can have logic which looks like:

```go
func appender(msg raftpb.Message) {
    var b pebble.Batch
    if len(msg.Entries) > 0 {
        b.Append(msg.Entries...)
    }
    if msg.HardState != nil {
        // NOTE: will change a bit with your comment above.
        b.SetHardState(msg.Entries...)
    }
    if len(msg.Responses) == 0 {
        b.CommitWithoutSync()
    } else {
        b.CommitWithSyncCallback(func() {
            sendMessages(msg.Responses...)
        })
    }
}
```
In a sense, we're no longer making a distinction about whether the "appender" logic needs to force an immediate sync or not. Instead, we're using the presence of response messages to decide whether to register an on-sync callback. The timing of that fsync can then be decided outside of this code.